pkg/cmd/roachtest/tests/cdc.go

```diff
@@ -1899,7 +1900,182 @@ func configureDBForMultiTablePTSBenchmark(db *gosql.DB) error {
+
+					}
+				}
+			}
+		}
+	}
+	return changefeedDistribution
+}
+
+func veryifyLeaseHolderDistribution(
+	db *gosql.DB, t test.Test, nodeToZone map[int]string,
+) map[string]int {
+	var rows *gosql.Rows
+	// Get lease holders for all ranges in tpcc database.
+	leaseHolderQuery := `SELECT r.start_pretty, r.replicas, r.replica_localities, r.lease_holder 
+	FROM crdb_internal.ranges r 
+	JOIN crdb_internal.tables t ON r.start_pretty like concat('/Table/', t.table_id::STRING,'%')
+	WHERE t.database_name = 'tpcc'`
+	rows, err := db.Query(leaseHolderQuery)
+	zoneToLeaseHolderCount := make(map[string]int)
+	require.NoError(t, err)
+	defer rows.Close()
+	for rows.Next() {
+		var startKeyPretty string
+		var replicas []uint8
+		var replicaLocalities []uint8
+		var leaseHolder int
+		require.NoError(t, rows.Scan(&startKeyPretty, &replicas, &replicaLocalities, &leaseHolder))
+		for indx := range replicas {
+			require.NotEqual(t, replicas[indx], 0)
+			replicas[indx]--
+		}
+		leaseHolder--
+		zoneToLeaseHolderCount[nodeToZone[leaseHolder]]++
+	}
+	return zoneToLeaseHolderCount
+}
+
 func registerCDC(r registry.Registry) {
+	r.Add(registry.TestSpec{
+		// This test
+		// 1. Creates a cluster with 3 nodes each in us-east and us-west
+		// 2. Runs a tpcc workload, then sets tpcc database to primary region us-west
+		// 3. Creates a changefeed with execution locality set to us-east
+		// 4. Gets the changefeed diagram and creates mappings
+
+		// This test is used to verify that ranges are evenly distributed across
+		// change aggregators in the execution_locality region while targeting tables
+		// whose primary region is different. In issue #2955, in that scenario,
+		// a single change aggregator (on the gateway node) would watch all the ranges.
+		// The above scenario occured with the older bin-packing oracle rather than
+		// the bulk oracle.
+		Name:             "cdc/multi-region-execution-locality-tpcc",
+		Owner:            registry.OwnerCDC,
+		Cluster:          r.MakeClusterSpec(7, spec.Geo(), spec.GatherCores(), spec.GCEZones("us-east1-b,us-west1-b")),
+		CompatibleClouds: registry.OnlyGCE,
+		Suites:           registry.Suites(),
+		Run: func(ctx context.Context, t test.Test, c cluster.Cluster) {
+			nodeToZone := map[int]string{
+				0: "us-east1-b",
+				1: "us-east1-b",
+				2: "us-east1-b",
+				3: "us-west1-b",
+				4: "us-west1-b",
+				5: "us-west1-b",
+			}
+			ct := newCDCTester(ctx, t, c)
+			defer ct.Close()
+
+			ct.runTPCCWorkload(tpccArgs{warehouses: 100})
+
+			var err error
+			_, err = ct.DB().Exec("ALTER DATABASE tpcc SET PRIMARY REGION 'us-west1'")
+			require.NoError(t, err)
+
+			feed := ct.newChangefeed(feedArgs{
+				sinkType: cloudStorageSink,
+				targets:  allTpccTargets,
+				opts: map[string]string{
+					"execution_locality": "'region=us-east1'",
+				},
+			})
+			ct.waitForWorkload()
+			feed.waitForCompletion()
```

Comment:
Did you try running this test? Seems to hang on this line, which seems expected given changefeeds don't complete unless they hit an error or they're initial-scan-only